lib/spack/spack/environment.py

```diff
@@ -1741,13 +1741,15 @@ def _read_lockfile_dict(self, d):
         root_hashes = set(self.concretized_order)
 
         specs_by_hash = {}
-        for dag_hash, node_dict in json_specs_by_hash.items():
-            specs_by_hash[dag_hash] = Spec.from_node_dict(node_dict)
+        for build_hash, node_dict in json_specs_by_hash.items():
+            spec = Spec.from_node_dict(node_dict)
+            setattr(spec, ht.build_hash.attr, build_hash)
```

Comment:
This likely deserves a comment so that people understand why envs were a special case.

Reply:
What is the situation with the full hash? It doesn't seem we have that too in the node dict and we don't store as a key. Should we start storing all the hashes as keys in the node dict?

Reply:
This fix is failing a unit test checking conversion from lockfile v1 to lockfile v2, so I think a quick solution is:
```diff
diff --git a/lib/spack/spack/environment.py b/lib/spack/spack/environment.py
index b678ae1943..fae23335fc 100644
--- a/lib/spack/spack/environment.py
+++ b/lib/spack/spack/environment.py
@@ -1744,9 +1744,10 @@ def _read_lockfile_dict(self, d):
         specs_by_hash = {}
         for build_hash, node_dict in json_specs_by_hash.items():
             spec = Spec.from_node_dict(node_dict)
-            # Build hash is stored as a key, but not as part of the node dict
-            # To ensure build hashes are not recomputed, we reattach here
-            setattr(spec, ht.build_hash.attr, build_hash)
+            if d['_meta']['lockfile-version'] > 1:
+                # Build hash is stored as a key, but not as part of the node dict
+                # To ensure build hashes are not recomputed, we reattach here
+                setattr(spec, ht.build_hash.attr, build_hash)
             specs_by_hash[build_hash] = spec
 
         for build_hash, node_dict in json_specs_by_hash.items():
```
Since we had a few issues with upgrading the format iof files n the past, we might want to investigate more structured solution to upgrading them later.

Reply:
> What is the situation with the full hash? It doesn't seem we have that too in the node dict and we don't store that as a key. Should we start storing all the hashes in the node dict?

I have an open PR (#25708) that does that.  I argue in the description that if we can't recompute hashes from concrete specs from yaml, then we need to write all the hashes to the lock.  That PR needs to be rebased now, since I started it before the spec yaml->json PR was merged.